pkg/txn/storage/tae/read.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2021 - 2022 Matrix Origin
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package taestorage
+
+import (
+	"bytes"
+	"context"
+	"encoding/gob"
+	apipb "github.com/matrixorigin/matrixone/pkg/pb/api"
+	"github.com/matrixorigin/matrixone/pkg/pb/txn"
+	"github.com/matrixorigin/matrixone/pkg/txn/storage"
+	"io"
+)
+
+// Read implements storage.TxnTAEStorage
+
+func (s *taeStorage) Read(
+	ctx context.Context,
+	txnMeta txn.TxnMeta,
+	op uint32,
+	payload []byte) (res storage.ReadResult, err error) {
+
+	switch op {
+
+	//case ...
+
+	case uint32(apipb.OpCode_OpGetLogTail):
+		return handleRead(
+			s, txnMeta, payload,
+			s.taeHandler.HandleGetLogTail,
+		)
+	}
+
+	return
+}
+
+func handleRead[Req any, Resp any](
+	s *taeStorage,
+	txnMeta txn.TxnMeta,
+	payload []byte,
+	fn func(
+		meta txn.TxnMeta,
+		req Req,
+		resp *Resp,
+	) (
+		err error,
+	),
+) (
+	res storage.ReadResult,
+	err error,
+) {
+
+	var req Req
+	if err := gob.NewDecoder(bytes.NewReader(payload)).Decode(&req); err != nil {
+		return nil, err
+	}
+
+	var resp Resp
+	defer logReq("read", req, txnMeta, &resp, &err)()
+	defer func() {
+		if closer, ok := (any)(resp).(io.Closer); ok {
```

Comment:
I will just put logutil ... stuff here, in one defer.

Reply:
> I wonder if you can try to get rid of TAE vector.
> 
> WHY do we need so many of that? Also, why do we need a proto vector? Isn't vector serialized and deserialized to bytes -- that is all?

`Proto vector` had been  involved in  before this PR .